content/general/concepts/what-is-the-prototype-pattern/what-is-the-prototype-pattern.md

```diff
@@ -0,0 +1,270 @@
+---
+Title: 'What is the Prototype Pattern?'
+Description: 'The _prototype pattern_ adds cloning functionality to an objets class. After the instantiation an initial object, the prototype 
+removes the need for the new keyword in creating subsequent objects at run time.'
+Subjects:
+- 'Computer Science'
+- 'Interview Prep'
+  Tags:
+- 'Classes'
+- 'Conceptual'
+- 'Constructors'
+- 'Objects'
+  CatalogContent:
+- 'learn-java'
+- 'paths/computer-science'
+---
+The _prototype pattern_ adds cloning functionality to an objects class. After the instantiation an initial object, the prototype removes the need for the `new` keyword in creating subsequent objects at run time. New objects are created using the property values of a pre-existing object.
+
+There are two types of copying associated with the prototype pattern. They are:
+- Shallow Copy (copies the immediate property values)
+- Deep Copy (copies the immediate values, plus any referenced object)
+
+## Example: Shallow Copy Vs Deep Copy
+To illustrate the difference between the shallow and deep copy, a model has been provided below, consisting of two classes `TvSeries` and `Episode`. Cloning functionality has been added to TvSeries by implementing the Cloneable interface.
+
+```java
+public class TvSeries implements Cloneable {
+  private final String name;
+  private final int seriesNo;
+  // Notice one or more properties are a complex type and mutable
+  private final List<Episode> episodes;
+
+  public TvSeries(String name, int seriesNo, List<Episode> episodes) {
+    this.name = name;
+    this.seriesNo = seriesNo;
+    this.episodes = episodes;
+  }
+    
+  // Shallow copy method
+  @Override
+  public TvSeries clone() throws CloneNotSupportedException {
+    // Cast to TvSeries as super returns Object
+    return (TvSeries) super.clone();
+  }
+
+  // Deep copy
+  public TvSeries deepCopy() {
+    // New Episode objects are created during the copy
+    return new TvSeries(this.name, this.seriesNo, this.episodes.stream()
+                                                               .map(episode -> new Episode(episode.getName(), episode.getEpNo()))
+                                                               .collect(Collectors.toList()));
+  }
+    
+  // Getters and toString
+}
+
+```
+
+```java
+public class Episode {
+  private String name;
+  private int epNo;
+
+  public Episode(String name, int epNo)
+  {
+    this.name = name;
+    this.epNo = epNo;
+  }
+
+  // Getters, setters and toString
+}
+```
+
+**Note:** The properties of the episode class are mutable. This is to illustrate a potential drawback in the shallow copy method implemented in the TvSeries class.
```

Comment:
This should explicitly mention the following code block illustrates the drawbacks.